text/0000-tracked-properties.md

```diff
@@ -227,34 +248,350 @@ memoization) if a property shows up as a bottleneck during profiling.
+the system. Tracked properties must also properly invalidate if they consume
+any of these properties.
+2. **Ember.set() directly**. If a tracked property accesses a simple, untracked
+property which is _later_ set with `Ember.set()`, it must have a way to know
+that this property has changed. This can occur in services, helper classes which
+have not yet converted to `@tracked`, and in many other edge cases.
+
+In the case of computed properties, we have all the information we need up front
+to allow tracked getters to watch for changes in the computed. They will notify
+whenever they are updated, so the same technique which is used for the autotrack
+stack can be used to watch them as well. Crucially, computed properties will
+always have a tag due to the fact that they are defined explicitly in class
+definitions, which means we have something to watch the very first time a
+computed property is accessed.
+
+In the edge case of non-computed, non-tracked property access, this is not the
+case. The property may be completely undecorated, meaning it may never have had
+a tag in the first place. With standard Javascript property access, we won't
+have an opportunity to _add_ a tag either. Autotrack will not work in these
+cases:
+
+```js
+const Config = Service.extend({
+  polling: {
+    shouldPoll: false,
+    pollInterval: -1,
+  },
+
+  init() {
+    this._super(...arguments);
+
+    fetch('config/api/url')
+      .then(r => r.json())
+      .then(polling => set(this, 'polling', polling));
+  }
+})
+
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    // When we access shouldPoll and pollInterval, they don't have tags or
+    // wrapped getters. We have no idea that they could update in the future.
+    let { shouldPoll, pollInterval } = this.config.polling;
+
+    return shouldPoll ? pollInterval : -1;
+  }
+}
+```
+
+To support this case, `Ember.get` will manually opt in to adding the tags for
+any properties which are accessed during a tracked getter. This allows tracked
+getters to safely access services and objects which are still using the legacy
+change tracking system:
+
+```js
+const Config = Service.extend({
+  polling: {
+    shouldPoll: false,
+    pollInterval: -1,
+  },
+
+  init() {
+    this._super(...arguments);
+
+    fetch('config/api/url')
+      .then(r => r.json())
+      .then(polling => set(this, 'polling', polling));
+  }
+})
+
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    // When we access shouldPoll and pollInterval, they don't have tags or
+    // wrapped getters. We have no idea that they could update in the future.
+    let shouldPoll = this.get('config.polling.shouldPoll');
+    let pollInterval = this.get('config.polling.pollInterval');
+
+    return shouldPoll ? pollInterval : -1;
+  }
+}
+```
+
+This solution will _only_ be necessary in cases where users do not know ahead of
```

Comment:
I think we should make it clear that this is only needed for addons that need to support older apps. The correct thing to do in the majority of cases is simply annotate the mutated, non-computed property with `@tracked`.

Because of this, I think we should present the `tracked()` functionality from below first, and present the `get()` workaround as an alternative for addon authors that need to handle third-party objects.